### PR TITLE
Add agent reconnects in the case of invalid token grants

### PR DESCRIPTION
**Pull request by nerdalert**
- Initializes a new OIDC client in the case where the token becomes invalid.
  This includes surpassing the keycloak offlineSessionIdleTimeout timer.
- If the onboard was via the one-time pass, exit with a message instead
  of prompting with a url in the logs.
- The offlineSessionIdleTimeout default value is 30 days, so it is not an
  error that will be triggered often for one-time joiners.
- In testing so far, the existing OIDC keycloak handles reconnects quite
  well as is, as long as offlineSessionIdleTimeout is not eclipsed.

### CODE REVIEW

**Comment by nerdalert** [comment]
The existing timeouts were a bit more durable than I was expecting. 

For example, I can:
- join with a one-time, 
- clip the connection to the api-server from the agent
- wait until the token is expiring > 10 minutes
- re-enable the agent connection
- watch the agent start getting peer listings again

I am focusing on one-time since creating a new client for the username/pass is trivial and doesn't affect anything but a new client for one-time forces a new URL visit. So I am going to let a disconnect go for a few hours and see what happens with a one-time login before I convert this from draft. 

Log output from a disconnect and then reconnect with this PR looks like the following where the scenario is [1] join [2] clip connectivity for the agent to the api-server [3] re-enable connectivity to the api-server:

- One-time code login

```
[07:13:18] (='o'=) [nexd]$ sudo ./nexd https://try.nexodus.local
Your device must be registered with Nexodus.
Your one-time code is: EXXP-TRUO
Please open the following URL in your browser to sign in:
https://auth.try.nexodus.local/realms/nexodus/device?user_code=EXXP-TRUO
Authentication succeeded.
{"level":"info","ts":1678407230.872014,"caller":"nexodus/keys.go:35","msg":"Existing key pair found at [ /usr/local/etc/wireguard/public.key ] and [ /usr/local/etc/wireguard/private.key ]"}
{"level":"info","ts":1678407230.936555,"caller":"nexodus/nexodus.go:230","msg":"Device belongs in organization: 4d5bf102-d00f-4f65-bb1f-8e80742e5917"}
{"level":"info","ts":1678407236.0915499,"caller":"nexodus/nexodus.go:299","msg":"Successfully registered device with UUID: 93ce8b28-26d2-40cc-b974-95fd51b21788"}
{"level":"info","ts":1678407236.141587,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.2] ] Peer Endpoint IP [ 192.168.1.201:51820 ] Peer Public Key [ chl2BGvvsgl2E+ARlbLlLN1o+fJy3xd+nb36TW6jo1Q= ] TunnelIP [ 100.100.0.2 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.141723,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.4] ] Peer Endpoint IP [ 192.168.1.201:51820 ] Peer Public Key [ LJAzqNs5RNt2yvwm8IE3Z8YTuvapqUrctgRauFanUWQ= ] TunnelIP [ 100.100.0.4 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.1417742,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.3] ] Peer Endpoint IP [ 192.168.1.201:51820 ] Peer Public Key [ S98S5D5CO/5zTJMvQJRqo4TpPV+fGgwZMM8axy90ZFQ= ] TunnelIP [ 100.100.0.3 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.141812,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.5] ] Peer Endpoint IP [ 192.168.1.201:51820 ] Peer Public Key [ Vk1bWLFxq50KvtOHZNmkEV32bauLsPy0KSH3XYQYcRw= ] TunnelIP [ 100.100.0.5 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.141929,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.6/32] ] Peer Endpoint IP [ 192.168.1.201:48630 ] Peer Public Key [ N07054ge4fRcwsfWMj4GLxGnObjvfo4fakvh1ILdLg4= ] TunnelIP [ 100.100.0.6 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.1419709,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.7/32] ] Peer Endpoint IP [ 192.168.1.201:43391 ] Peer Public Key [ KMeeusqr5PZSrQDwPt3DoKXhl54dayld4roaIjR65g0= ] TunnelIP [ 100.100.0.7 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.142005,"caller":"nexodus/wg-peers.go:130","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.8/32] ] Peer Endpoint IP [ 192.168.1.201:37562 ] Peer Public Key [ tg/cNbk/hzwhl5T19bN0lvdhoJoQcvUgatbDFfhWa08= ] TunnelIP [ 100.100.0.8 ] Organization [ 4d5bf102-d00f-4f65-bb1f-8e80742e5917 ]"}
{"level":"info","ts":1678407236.1423979,"caller":"nexodus/wg-peers.go:173","msg":"New local Wireguard interface address assigned: 100.100.0.1"}
{"level":"error","ts":1678407291.060649,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/4d5bf102-d00f-4f65-bb1f-8e80742e5917/devices\": read tcp 192.168.1.9:50760->54.87.252.157:443: read: operation timed out"}
{"level":"error","ts":1678407296.064272,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678407301.069468,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/4d5bf102-d00f-4f65-bb1f-8e80742e5917/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678407306.07473,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678407311.078086,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/4d5bf102-d00f-4f65-bb1f-8e80742e5917/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678407316.0822709,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678407321.088298,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/4d5bf102-d00f-4f65-bb1f-8e80742e5917/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
Your device must be registered with Nexodus.
Your one-time code is: BNMV-IZNR
Please open the following URL in your browser to sign in:
https://auth.try.nexodus.local/realms/nexodus/device?user_code=BNMV-IZNR
Authentication succeeded.
{"level":"info","ts":1678407356.984885,"caller":"nexodus/nexodus.go:347","msg":"Nexodus agent has re-established a connection to the api-server"}
```



- u/pass login output

```
[07:38:41] (='o'=) [nexd]$ sudo ./nexd --username kitteh1 --password floofykittens https://try.nexodus.local
Password:
{"level":"info","ts":1678408742.016913,"caller":"nexodus/keys.go:35","msg":"Existing key pair found at [ /usr/local/etc/wireguard/public.key ] and [ /usr/local/etc/wireguard/private.key ]"}
{"level":"info","ts":1678408742.068675,"caller":"nexodus/nexodus.go:230","msg":"Device belongs in organization: 9098731c-f256-4f91-bdad-dde582721501"}
{"level":"info","ts":1678408747.1308749,"caller":"nexodus/nexodus.go:299","msg":"Successfully registered device with UUID: ba35ca6b-0ab9-4565-a453-e735fb167cb4"}
{"level":"info","ts":1678408747.177282,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.7/32] ] Peer Endpoint IP [ 174.129.58.36:46887 ] Peer Public Key [ QXExSy5CTiKgvdKeMwZxS4TFqsWK5mtMqadZew8g8AI= ] TunnelIP [ 100.100.0.7 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.1783319,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.3/32] ] Peer Endpoint IP [ 54.167.97.135:40542 ] Peer Public Key [ fGjIwqk8rzvuxqEZVghrvDbKU1wKpR2pNFkD2++UXG0= ] TunnelIP [ 100.100.0.3 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.178344,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.5/32] ] Peer Endpoint IP [ 54.227.102.183:52767 ] Peer Public Key [ 6/CwH/gzz9jdKnxeVpWFivycFQMBkniLeBNFQq0+f04= ] TunnelIP [ 100.100.0.5 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.178355,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.10/32] ] Peer Endpoint IP [ 54.165.242.46:33551 ] Peer Public Key [ ymN5YXXxRdVnpQF7dfAdNHdaIzwM7c4CsEdTJvKZ8E0= ] TunnelIP [ 100.100.0.10 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.178432,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.6/32] ] Peer Endpoint IP [ 35.173.211.114:40322 ] Peer Public Key [ FNHhpJpeQcpKC+QCLHqwX9CahOo/PQEG8AWNj9RF9lU= ] TunnelIP [ 100.100.0.6 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.178441,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.8/32] ] Peer Endpoint IP [ 44.211.198.0:50039 ] Peer Public Key [ itc587APYo5VxKmjBRZjh8LCJSFDgM95PfDUNrcxsmo= ] TunnelIP [ 100.100.0.8 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.178453,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.9/32] ] Peer Endpoint IP [ 44.202.61.34:39506 ] Peer Public Key [ C+7gYhnz3rXT06I/Ga0ZFZdSD5FzLcd91CuZTRdC0yk= ] TunnelIP [ 100.100.0.9 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.1785321,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.2/32] ] Peer Endpoint IP [ 50.17.69.84:51169 ] Peer Public Key [ d8aGJlGvHu2YAURbbA19WM+PfNO5Fk6692C22V6aWC0= ] TunnelIP [ 100.100.0.2 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.1785421,"caller":"nexodus/wg-peers.go:152","msg":"Peer Configuration - Peer AllowedIPs [ [100.100.0.4/32] ] Peer Endpoint IP [ 3.88.203.9:56543 ] Peer Public Key [ 7mo13J8xt47YD0jAyVR4g0dAfaPfU33ZHiS4KJhesQQ= ] TunnelIP [ 100.100.0.4 ] Organization [ 9098731c-f256-4f91-bdad-dde582721501 ]"}
{"level":"info","ts":1678408747.178551,"caller":"nexodus/wg-peers.go:173","msg":"New local Wireguard interface address assigned: 100.100.0.11"}
{"level":"error","ts":1678408822.4539819,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/9098731c-f256-4f91-bdad-dde582721501/devices\": read tcp 192.168.1.9:53176->54.87.252.157:443: read: can't assign requested address"}
{"level":"error","ts":1678408827.463628,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408832.475053,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/9098731c-f256-4f91-bdad-dde582721501/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408837.495771,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408842.5030348,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/9098731c-f256-4f91-bdad-dde582721501/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408847.513758,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408852.5239522,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/9098731c-f256-4f91-bdad-dde582721501/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408857.525942,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconnect to the api-server, retrying in 5s seconds: Post \"https://api.try.nexodus.local/login/start\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"error","ts":1678408862.528348,"caller":"nexodus/nexodus.go:338","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/9098731c-f256-4f91-bdad-dde582721501/devices\": dial tcp 54.87.252.157:443: connect: network is unreachable"}
{"level":"info","ts":1678408873.217678,"caller":"nexodus/nexodus.go:347","msg":"Nexodus agent has re-established a connection to the api-server"}
```

**Comment by russellb** [comment]
Issue #573

**Comment by nerdalert** [comment]
Updated git comments. After more testing, we currently appear to handle disconnects better than I first thought 🎉  What we don't handle is the token becoming invalid. This patch adds that capability and was tested by dialing down the keycloak timers to get the returned error to handle.

One caveat in the patch is `errors.Is` does not appear to unwrap the error message being returned. I'm at diminishing returns on it and simply using `if strings.Contains(err.Error(), invalidTokenGrant.Error()) {` instead here ¯\_(ツ)_/¯ 

```go
if errors.Is(err, invalidTokenGrant) {
    ...
}
```

- u/pass

```
{"level":"error","ts":1678479641.750132,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/c0c75ff6-fd64-4245-8630-ecc0999e5ce0/devices\": oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Offline session not active\"}"}
{"level":"info","ts":1678479643.7441978,"caller":"nexodus/nexodus.go:356","msg":"Nexodus agent has re-established a connection to the api-server"}
```

- one-time
```
{"level":"error","ts":1678479323.535163,"caller":"nexodus/nexodus.go:343","msg":"Failed to reconcile state with the nexodus API server: Get \"https://api.try.nexodus.local/api/organizations/c0c75ff6-fd64-4245-8630-ecc0999e5ce0/devices\": oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Offline user session not found\"}"}
{"level":"fatal","ts":1678479323.535261,"caller":"nexodus/nexodus.go:359","msg":"The token grant has expired due to an extended period offline, please restart the agent for a one-time auth or login with --username --password to automatically reconenct"}
```

File: internal/nexodus/nexodus.go [hunk 1]
```diff
@@ -335,8 +340,25 @@ func (ax *Nexodus) Start(ctx context.Context, wg *sync.WaitGroup) error {
...
+						if err != nil {
+							ax.logger.Errorf("Failed to reconnect to the api-server, retrying in %v seconds: %v", pollInterval, err)
+						} else {
+							ax.client = c
+							ax.SetStatus(NexdStatusRunning, "")
+							ax.logger.Infoln("Nexodus agent has re-established a connection to the api-server")
+						}
+					} else {
+						ax.logger.Fatalf("The token grant has expired due to an extended period offline, please " +
+							"restart the agent for a one-time auth or login with --username --password to automatically reconenct")
```
Comment (vishnoianil) [inline internal/nexodus/nexodus.go:359]: nit pick : typo: "reconnect"

File: internal/nexodus/nexodus.go [hunk 2]
```diff
@@ -335,8 +340,25 @@ func (ax *Nexodus) Start(ctx context.Context, wg *sync.WaitGroup) error {
...
-				// to handle disconnects and/or timeouts etc...
 				ax.logger.Errorf("Failed to reconcile state with the nexodus API server: %v", err)
+				// if the token grant becomes invalid expires refresh or exit depending on the onboard method
+				if strings.Contains(err.Error(), invalidTokenGrant.Error()) {
+					if ax.username != "" {
+						c, err := client.NewClient(ctx, ax.controllerURL.String(), func(msg string) {
+							ax.SetStatus(NexdStatusAuth, msg)
+						}, option)
+						if err != nil {
+							ax.logger.Errorf("Failed to reconnect to the api-server, retrying in %v seconds: %v", pollInterval, err)
```
Comment (vishnoianil) [inline internal/nexodus/nexodus.go:351]: Just a suggestion : I think it's good to probably slow down the retry attempt with every failed attempt

**Review by vishnoianil — COMMENTED** [review]
Looks good to me, minor comments.

File: internal/nexodus/nexodus.go (hunk 2 above)
Comment (nerdalert) [reply internal/nexodus/nexodus.go:351]: Agree! I was thinking the whole `util.Waitgroup` should have a backoff though, that captures the other reconcile group for the discovery node as well to relieve unnecessary pressure on the controller. This retry will only be hit like once a month as far as I can tell with the Keycloack defaults. We need to capture other errors tbd as well.

Comment (vishnoianil) [reply internal/nexodus/nexodus.go:351]: I think once in month probe doesn't need special treatment as of now. Sounds good to me.

**Review by vishnoianil — APPROVED** [review]

**Comment by vishnoianil** [comment]
PR looks good to me, I approved it, but holding the merge, in case other reviewers want to take a look (given that we had long discussion on it over the slack).

**Review by chirino — APPROVED** [review]
lgtm

**Review by vishnoianil — APPROVED** [review]

**Pull request merged**